moo.js

```diff
@@ -469,10 +433,55 @@
   }
 
 
+  function keywords(rule, namer, keywordList) {
+    if (typeof namer !== 'function') {
+      var type = namer
+      namer = function() { return type }
+    }
+
+    function str(x) { return JSON.stringify(x) }
+
+    var byLength = Object.create(null)
+    keywordList.forEach(function(keyword) {
+      (byLength[keyword.length] = byLength[keyword.length] || []).push(keyword)
+    })
+
+    // fast string lookup
```

Comment:
Not personally, but [I've read](http://marijnhaverbeke.nl/blog/acorn.html) that switching on the length first is a helpful thing to do.